File: engine_py/tests/test_map_api.py
def test_map_api(make_world):
    world = make_world()
    world.add_cell(0, 0, 0)
    world.add_cell(1, 0, 0)
    world.add_cell(0, 1, 0)

    # Add neighbors explicitly
    world.add_neighbor((0, 0, 0), (1, 0, 0))
    world.add_neighbor((0, 0, 0), (0, 1, 0))

    topo = world.get_map_topology_type()
    assert topo == "square"

    cells = world.get_all_cells()
    assert len(cells) >= 3

    cell = {"Square": {"x": 0, "y": 0, "z": 0}}
    neighbors = world.get_neighbors(cell)
    assert len(neighbors) > 0


def test_entities_in_zlevel_returns_matching_entities(make_world):
    world = make_world()
    world.add_cell(2, 0, 1)

    lower = world.spawn_entity()
    world.set_component(lower, "Position", {"pos": {"Square": {"x": 0, "y": 0, "z": 0}}})
    upper = world.spawn_entity()
    world.set_component(upper, "Position", {"pos": {"Square": {"x": 2, "y": 0, "z": 1}}})

    on_zero = world.entities_in_zlevel(0)
    assert lower in on_zero, "Lower entity should be on z-level 0"
    assert upper not in on_zero, "Upper entity should not be on z-level 0"

    on_one = world.entities_in_zlevel(1)
    assert upper in on_one, "Upper entity should be on z-level 1"
    assert lower not in on_one, "Lower entity should not be on z-level 1"


def test_move_entity_3d_changes_z(make_world):
    world = make_world()
    world.add_cell(0, 0, 0)
    world.add_cell(0, 0, 2)

    eid = world.spawn_entity()
    world.set_component(eid, "Position", {"pos": {"Square": {"x": 0, "y": 0, "z": 0}}})

    world.move_entity_3d(eid, 0, 0, 2)

    pos = world.get_component(eid, "Position")
    assert pos["pos"]["Square"]["x"] == 0, "x should be unchanged by vertical movement"
    assert pos["pos"]["Square"]["y"] == 0, "y should be unchanged by vertical movement"
    assert pos["pos"]["Square"]["z"] == 2, "z should increase by dz"

    assert eid in world.entities_in_zlevel(2), "Entity should be found on new z-level"
    assert eid not in world.entities_in_zlevel(0), "Entity should leave old z-level"


def test_move_entity_3d_shifts_xy_and_z(make_world):
    world = make_world()

    eid = world.spawn_entity()
    world.set_component(eid, "Position", {"pos": {"Square": {"x": 1, "y": 1, "z": 0}}})

    world.move_entity_3d(eid, 2, 3, 1)

    pos = world.get_component(eid, "Position")
    assert pos["pos"]["Square"]["x"] == 3, "x should shift by dx"
    assert pos["pos"]["Square"]["y"] == 4, "y should shift by dy"
    assert pos["pos"]["Square"]["z"] == 1, "z should shift by dz"

File: engine/scripts/lua/tests/test_map_api.lua
local assert = require("assert")

local function contains(arr, value)
	for _, v in ipairs(arr) do
		if v == value then
			return true
		end
	end
	return false
end

local function test_map_api()
	add_cell(0, 0, 0)
	add_cell(1, 0, 0)
	add_cell(0, 1, 0)

	-- Add neighbors explicitly
	add_neighbor({ 0, 0, 0 }, { 1, 0, 0 })
	add_neighbor({ 0, 0, 0 }, { 0, 1, 0 })

	local topo = get_map_topology_type()
	assert.equals(topo, "square", "Topology should be square")

	local cells = get_all_cells()
	assert.is_true(#cells >= 3, "Should have at least 3 cells")

	local cell = { Square = { x = 0, y = 0, z = 0 } }
	local neighbors = get_neighbors(cell)
	assert.is_true(#neighbors > 0, "Cell should have neighbors")
end

local function test_entities_in_zlevel_returns_matching_entities()
	add_cell(2, 0, 1)

	local lower = spawn_entity()
	set_component(lower, "Position", { pos = { Square = { x = 0, y = 0, z = 0 } } })
	local upper = spawn_entity()
	set_component(upper, "Position", { pos = { Square = { x = 2, y = 0, z = 1 } } })

	local on_zero = entities_in_zlevel(0)
	assert.is_true(contains(on_zero, lower), "Lower entity should be on z-level 0")
	assert.is_false(contains(on_zero, upper), "Upper entity should not be on z-level 0")

	local on_one = entities_in_zlevel(1)
	assert.is_true(contains(on_one, upper), "Upper entity should be on z-level 1")
	assert.is_false(contains(on_one, lower), "Lower entity should not be on z-level 1")
end

local function test_move_entity_3d_changes_z()
	add_cell(0, 0, 0)
	add_cell(0, 0, 2)

	local eid = spawn_entity()
	set_component(eid, "Position", { pos = { Square = { x = 0, y = 0, z = 0 } } })

	move_entity_3d(eid, 0, 0, 2)

	local pos = get_component(eid, "Position")
	assert.equals(pos.pos.Square.x, 0, "x should be unchanged by vertical movement")
	assert.equals(pos.pos.Square.y, 0, "y should be unchanged by vertical movement")
	assert.equals(pos.pos.Square.z, 2, "z should increase by dz")

	assert.is_true(contains(entities_in_zlevel(2), eid), "Entity should be found on new z-level")
	assert.is_false(contains(entities_in_zlevel(0), eid), "Entity should leave old z-level")
end

local function test_move_entity_3d_shifts_xy_and_z()
	local eid = spawn_entity()
	set_component(eid, "Position", { pos = { Square = { x = 1, y = 1, z = 0 } } })

	move_entity_3d(eid, 2, 3, 1)

	local pos = get_component(eid, "Position")
	assert.equals(pos.pos.Square.x, 3, "x should shift by dx")
	assert.equals(pos.pos.Square.y, 4, "y should shift by dy")
	assert.equals(pos.pos.Square.z, 1, "z should shift by dz")
end

return {
	test_map_api = test_map_api,
	test_entities_in_zlevel_returns_matching_entities = test_entities_in_zlevel_returns_matching_entities,
	test_move_entity_3d_changes_z = test_move_entity_3d_changes_z,
	test_move_entity_3d_shifts_xy_and_z = test_move_entity_3d_shifts_xy_and_z,
}
